### PR TITLE
🐛 Make placement translator react more fully.

### LINKED ISSUE #320: bug: placement-translator doesn't respond to updating/deleting of edgeplacements as expected
### Describe the bug

It seems that placement-translator doesn't respond to updating/deleting of edgeplacements as expected.

I found this issue when I was trying to setup the Turbo demo using manifests living [here](https://github.com/edge-experiments/gitops-source/tree/main/edge-mc), where I need to change the [locationSelectors](https://github.com/edge-experiments/gitops-source/blob/768059253a942054ec0a7e75f856f591a43cf71c/edge-mc/placements/edgeplacements.yaml#L9) of an edgeplacement, and expect the workload to be moved from one location to another.

To verify the issue, I reproduce it using documented steps in [example1.md](https://github.com/kcp-dev/edge-mc/blob/main/docs/content/en/docs/Coding%20Milestones/PoC2023q1/example1.md) and a few extra steps.

### Environment to reproduce
kcp v0.11.0
edge-mc d6453e56ebe647026d5f3fb39ac35a22bf22002b

```console
ubuntu@ip-172-31-31-125:~/kcp$ git log -1
commit 771cfc8ac15407290c3a4f50a0d65d2511ad0cb6 (HEAD, tag: v0.11.0, tag: pkg/apis/v0.11.0)
Merge: 64341429 e5445582
Author: OpenShift Merge Robot <openshift-merge-robot@users.noreply.github.com>
Date:   Wed Feb 22 17:38:27 2023 +0100

    Merge pull request #2842 from ncdc/no-direct-klog
    
    🌱 Remove direct klog usage
ubuntu@ip-172-31-31-125:~/kcp$ ./bin/kcp start &> kcp-output.log &
[1] 690515
ubuntu@ip-172-31-31-125:~/kcp$ tail -1 kcp-output.log 
I0420 03:17:37.068887  690515 replication_reconcile.go:178] "Creating object in global cache" reconciler="kcp-replication-controller" key="v1.clusterrolebindings.rbac.authorization.k8s.io::root|system:kcp:authenticated:apiexport:tenancy:bind" reconcilerKey="root|system:kcp:authenticated:apiexport:tenancy:bind"
ubuntu@ip-172-31-31-125:~/kcp$ kc ws .
Current workspace is "root".
ubuntu@ip-172-31-31-125:~/kcp$ cd ../edge-mc/
ubuntu@ip-172-31-31-125:~/edge-mc$ git log -1
commit d6453e56ebe647026d5f3fb39ac35a22bf22002b (HEAD -> main, origin/main)
Merge: 282c7ce 7a153bf
Author: OpenShift Merge Robot <openshift-merge-robot@users.noreply.github.com>
Date:   Tue Apr 18 19:24:08 2023 -0400

    Merge pull request #315 from MikeSpreitzer/add-ls-demo
    
    ✨ Add cmd/ls-syncer-config
```

### Steps documented in 'example1'

This section is long. But it's simply following the steps in [example1.md](https://github.com/kcp-dev/edge-mc/blob/main/docs/content/en/docs/Coding%20Milestones/PoC2023q1/example1.md), and everything goes fine as documented.
Please feel free to skip to next section.

```console
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root
Current workspace is "root".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws create imw-1 --enter
Workspace "imw-1" (type root:organization) created. Waiting for it to be ready...
Workspace "imw-1" (type root:organization) is ready to use.
Current workspace is "root:imw-1" (type root:organization).
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: workload.kcp.io/v1alpha1
kind: SyncTarget
metadata:
  name: sync-target-f
  labels:
    example: si
    extended: non
spec:
  cells:
    foo: bar
EOF
synctarget.workload.kcp.io/sync-target-f created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: scheduling.kcp.io/v1alpha1
kind: Location
metadata:
  name: location-f
  labels:
    env: prod
spec:
  resource: {group: workload.kcp.io, version: v1alpha1, resource: synctargets}
  instanceSelector:
    matchLabels: {"example":"si", "extended":"non"}
EOF
location.scheduling.kcp.io/location-f created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl delete locations default
location.scheduling.kcp.io "default" deleted
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: workload.kcp.io/v1alpha1
kind: SyncTarget
metadata:
  name: sync-target-g
  labels:
    example: si
    extended: si
spec:
  cells:
    bar: baz
EOF
synctarget.workload.kcp.io/sync-target-g created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: scheduling.kcp.io/v1alpha1
kind: Location
metadata:
  name: location-g
  labels:
    env: prod
    extended: si
spec:
  resource: {group: workload.kcp.io, version: v1alpha1, resource: synctargets}
  instanceSelector:
    matchLabels: {"example":"si", "extended":"si"}
EOF
location.scheduling.kcp.io/location-g created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root
Current workspace is "root".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws create espw --enter
Workspace "espw" (type root:organization) created. Waiting for it to be ready...
Workspace "espw" (type root:organization) is ready to use.
Current workspace is "root:espw" (type root:organization).
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl create -f config/crds
customresourcedefinition.apiextensions.k8s.io/customizers.edge.kcp.io created
customresourcedefinition.apiextensions.k8s.io/edgeplacements.edge.kcp.io created
customresourcedefinition.apiextensions.k8s.io/edgesyncconfigs.edge.kcp.io created
customresourcedefinition.apiextensions.k8s.io/singleplacementslices.edge.kcp.io created
customresourcedefinition.apiextensions.k8s.io/syncerconfigs.edge.kcp.io created
customresourcedefinition.apiextensions.k8s.io/apiresources.meta.kcp.io created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl create -f config/exports
apiexport.apis.kcp.io/edge.kcp.io created
apiexport.apis.kcp.io/meta.kcp.io created
apiresourceschema.apis.kcp.io/v230220-d814bdd.apiresources.meta.kcp.io created
apiresourceschema.apis.kcp.io/v230227-655d6c9.customizers.edge.kcp.io created
apiresourceschema.apis.kcp.io/v230317-d43d2d5.edgeplacements.edge.kcp.io created
apiresourceschema.apis.kcp.io/v230413-5a4d4ba.edgesyncconfigs.edge.kcp.io created
apiresourceschema.apis.kcp.io/v230303-6f7f2fe.singleplacementslices.edge.kcp.io created
apiresourceschema.apis.kcp.io/v230407-c87b2365.syncerconfigs.edge.kcp.io created
ubuntu@ip-172-31-31-125:~/edge-mc$ ./bin/mailbox-controller -v 2 &> mailbox-controller.log &
[2] 690982
ubuntu@ip-172-31-31-125:~/edge-mc$ tail -1 mailbox-controller.log 
I0420 03:21:21.177891  690982 controller.go:299] "Created APIBinding" worker=0 mbwsName="1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631" mbwsCluster="pgy6ew6j1mz900sj" bindingName="bind-edge" resourceVersion="911"
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get Workspaces
NAME                                                       TYPE        REGION   PHASE   URL                                                    AGE
1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c   universal            Ready   https://172.31.31.125:6443/clusters/bpz9irwdv41kef7b   52s
1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631   universal            Ready   https://172.31.31.125:6443/clusters/pgy6ew6j1mz900sj   52s
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root
Current workspace is "root".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws create my-org --enter
Workspace "my-org" (type root:organization) created. Waiting for it to be ready...
Workspace "my-org" (type root:organization) is ready to use.
Current workspace is "root:my-org" (type root:organization).
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws create wmw-c --enter
Workspace "wmw-c" (type root:universal) created. Waiting for it to be ready...
Workspace "wmw-c" (type root:universal) is ready to use.
Current workspace is "root:my-org:wmw-c" (type root:universal).
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: apis.kcp.io/v1alpha1
kind: APIBinding
metadata:
  name: bind-kube
spec:
  reference:
    export:
      path: "root:compute"
      name: kubernetes
EOF
apibinding.apis.kcp.io/bind-kube created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: commonstuff
  labels: {common: "si"}
---
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: commonstuff
  name: httpd-htdocs
data:
  index.html: |
    <!DOCTYPE html>
    <html>
      <body>
        This is a common web site.
      </body>
    </html>
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: commonstuff
  name: commond
spec:
  selector: {matchLabels: {app: common} }
  template:
    metadata:
      labels: {app: common}
    spec:
      containers:
      - name: httpd
        image: library/httpd:2.4
        ports:
        - name: http
          containerPort: 80
          hostPort: 8081
          protocol: TCP
        volumeMounts:
        - name: htdocs
          readOnly: true
          mountPath: /usr/local/apache2/htdocs
      volumes:
      - name: htdocs
        configMap:
          name: httpd-htdocs
          optional: false
EOF
namespace/commonstuff created
configmap/httpd-htdocs created
deployment.apps/commond created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: apis.kcp.io/v1alpha1
kind: APIBinding
metadata:
  name: bind-espw
spec:
  reference:
    export:
      path: root:espw
      name: edge.kcp.io
EOF
apibinding.apis.kcp.io/bind-espw created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: edge.kcp.io/v1alpha1
kind: EdgePlacement
metadata:
  name: edge-placement-c
spec:
  locationSelectors:
  - matchLabels: {"env":"prod"}
  namespaceSelector:
    matchLabels: {"common":"si"}
  nonNamespacedObjects:
  - apiGroup: apis.kcp.io
    resources: [ "apibindings" ]
    resourceNames: [ "bind-kube" ]
  upsync:
  - apiGroup: "group1.test"
    resources: ["sprockets", "flanges"]
    namespaces: ["orbital"]
    names: ["george", "cosmo"]
  - apiGroup: "group2.test"
    resources: ["cogs"]
    names: ["william"]
EOF
edgeplacement.edge.kcp.io/edge-placement-c created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:my-org
Current workspace is "root:my-org".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws create wmw-s --enter
Workspace "wmw-s" (type root:universal) created. Waiting for it to be ready...
Workspace "wmw-s" (type root:universal) is ready to use.
Current workspace is "root:my-org:wmw-s" (type root:universal).
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: apis.kcp.io/v1alpha1
kind: APIBinding
metadata:
  name: bind-kube
spec:
  reference:
    export:
      path: "root:compute"
      name: kubernetes
EOF
apibinding.apis.kcp.io/bind-kube created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: v1
kind: Namespace
metadata:
  name: specialstuff
  labels: {special: "si"}
---
apiVersion: v1
kind: ConfigMap
metadata:
  namespace: specialstuff
  name: httpd-htdocs
data:
  index.html: |
    <!DOCTYPE html>
    <html>
      <body>
        This is a special web site.
      </body>
    </html>
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: specialstuff
  name: speciald
spec:
  selector: {matchLabels: {app: special} }
  template:
    metadata:
      labels: {app: special}
    spec:
      containers:
      - name: httpd
        image: library/httpd:2.4
        ports:
        - name: http
          containerPort: 80
          hostPort: 8082
          protocol: TCP
        volumeMounts:
        - name: htdocs
          readOnly: true
          mountPath: /usr/local/apache2/htdocs
      volumes:
      - name: htdocs
        configMap:
          name: httpd-htdocs
          optional: false
EOF
namespace/specialstuff created
configmap/httpd-htdocs created
deployment.apps/speciald created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: apis.kcp.io/v1alpha1
kind: APIBinding
metadata:
  name: bind-espw
spec:
  reference:
    export:
      path: root:espw
      name: edge.kcp.io
EOF
apibinding.apis.kcp.io/bind-espw created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl apply -f - <<EOF
apiVersion: edge.kcp.io/v1alpha1
kind: EdgePlacement
metadata:
  name: edge-placement-s
spec:
  locationSelectors:
  - matchLabels: {"env":"prod","extended":"si"}
  namespaceSelector: 
    matchLabels: {"special":"si"}
  nonNamespacedObjects:
  - apiGroup: apis.kcp.io
    resources: [ "apibindings" ]
    resourceNames: [ "bind-kube" ]
  upsync:
  - apiGroup: "group1.test"
    resources: ["sprockets", "flanges"]
    namespaces: ["orbital"]
    names: ["george", "cosmo"]
  - apiGroup: "group3.test"
    resources: ["widgets"]
    names: ["*"]
EOF
edgeplacement.edge.kcp.io/edge-placement-s created
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:espw
Current workspace is "root:espw".
ubuntu@ip-172-31-31-125:~/edge-mc$ ./bin/scheduler &> scheduler.log &
[3] 691284
ubuntu@ip-172-31-31-125:~/edge-mc$ tail -1 scheduler.log 
I0420 03:26:09.056740  691284 reconcile_on_synctarget.go:182] "updated SinglePlacementSlice" controller="edge-scheduler" triggeringKind=SyncTarget key="1gcc1zbbgm94lod9|sync-target-f" syncTargetWorkspace="1gcc1zbbgm94lod9" syncTarget="sync-target-f" workloadWorkspace="18hvxcwlfhako08h" singlePlacementSlice="edge-placement-c"
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:my-org:wmw-c
Current workspace is "root:my-org:wmw-c".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get SinglePlacementSlice -o yaml
apiVersion: v1
items:
- apiVersion: edge.kcp.io/v1alpha1
  destinations:
  - cluster: 1gcc1zbbgm94lod9
    locationName: location-g
    syncTargetName: sync-target-g
    syncTargetUID: ca3eb1cd-a8ac-4b4a-85b7-0be352500631
  - cluster: 1gcc1zbbgm94lod9
    locationName: location-f
    syncTargetName: sync-target-f
    syncTargetUID: a113b61e-83ed-4eac-86cc-73f114d79e4c
  kind: SinglePlacementSlice
  metadata:
    annotations:
      kcp.io/cluster: 18hvxcwlfhako08h
    creationTimestamp: "2023-04-20T03:26:09Z"
    generation: 2
    name: edge-placement-c
    ownerReferences:
    - apiVersion: edge.kcp.io/v1alpha1
      kind: EdgePlacement
      name: edge-placement-c
      uid: 7b7ea4f0-5766-4e86-904f-aa18009d57c9
    resourceVersion: "1258"
    uid: 7af8c851-7963-427b-b3e6-2800f4912a34
kind: List
metadata:
  resourceVersion: ""
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:my-org:wmw-s
Current workspace is "root:my-org:wmw-s".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get SinglePlacementSlice -o yaml
apiVersion: v1
items:
- apiVersion: edge.kcp.io/v1alpha1
  destinations:
  - cluster: 1gcc1zbbgm94lod9
    locationName: location-g
    syncTargetName: sync-target-g
    syncTargetUID: ca3eb1cd-a8ac-4b4a-85b7-0be352500631
  kind: SinglePlacementSlice
  metadata:
    annotations:
      kcp.io/cluster: 23qa6hllb50lvxjd
    creationTimestamp: "2023-04-20T03:26:08Z"
    generation: 1
    name: edge-placement-s
    ownerReferences:
    - apiVersion: edge.kcp.io/v1alpha1
      kind: EdgePlacement
      name: edge-placement-s
      uid: 4d015af9-61af-4a8f-b052-525e8ca777cd
    resourceVersion: "1255"
    uid: 781e5f45-f7da-4c0e-9da9-20bd81c93aa4
kind: List
metadata:
  resourceVersion: ""
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:espw
Current workspace is "root:espw".
ubuntu@ip-172-31-31-125:~/edge-mc$ ./bin/placement-translator &> placement-translator.log &
[4] 691389
ubuntu@ip-172-31-31-125:~/edge-mc$ date
Thu Apr 20 03:28:36 UTC 2023
ubuntu@ip-172-31-31-125:~/edge-mc$ date
Thu Apr 20 03:29:18 UTC 2023
ubuntu@ip-172-31-31-125:~/edge-mc$ date
Thu Apr 20 03:29:45 UTC 2023
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c
Current workspace is "root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c".
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get SyncerConfig the-one -o yaml
apiVersion: edge.kcp.io/v1alpha1
kind: SyncerConfig
metadata:
  annotations:
    kcp.io/cluster: bpz9irwdv41kef7b
  creationTimestamp: "2023-04-20T03:27:21Z"
  generation: 3
  name: the-one
  resourceVersion: "1263"
  uid: 7f976b91-2728-4c8f-83bf-06bcf3a6f642
spec:
  namespaceScope:
    namespaces:
    - commonstuff
    resources:
    - apiVersion: v1
      group: ""
      resource: secrets
    - apiVersion: v1
      group: ""
      resource: configmaps
    - apiVersion: v1
      group: ""
      resource: services
    - apiVersion: v1
      group: ""
      resource: pods
    - apiVersion: v1
      group: coordination.k8s.io
      resource: leases
    - apiVersion: v1
      group: apps
      resource: deployments
    - apiVersion: v1
      group: networking.k8s.io
      resource: ingresses
    - apiVersion: v1
      group: rbac.authorization.k8s.io
      resource: roles
    - apiVersion: v1
      group: ""
      resource: resourcequotas
    - apiVersion: v1
      group: ""
      resource: serviceaccounts
    - apiVersion: v1
      group: ""
      resource: limitranges
    - apiVersion: v1
      group: rbac.authorization.k8s.io
      resource: rolebindings
  upsync:
  - apiGroup: group1.test
    names:
    - george
    - cosmo
    namespaces:
    - orbital
    resources:
    - sprockets
    - flanges
  - apiGroup: group2.test
    names:
    - william
    resources:
    - cogs
status: {}
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME          STATUS   AGE
commonstuff   Active   2m57s
default       Active   8m58s
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get deployments -A
NAMESPACE     NAME      READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff   commond   0/0     0            0           3m17s
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:espw:1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631
Current workspace is "root:espw:1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631".
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME           STATUS   AGE
commonstuff    Active   4m9s
default        Active   10m
specialstuff   Active   4m9s
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl get deployments -A
NAMESPACE      NAME       READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff    commond    0/0     0            0           4m11s
specialstuff   speciald   0/0     0            0           4m11s
```

#### Update an EdgePlacement
In these steps, I `kubectl edit` `edge-placement-s` to remove `"extended":"si"` in `locationSelectors`.
By doing so, `edge-placement-s` is not 'special' any more, it selects the same set of locations with `edge-placement-c`.
Therefore, it's expected that 'specialstuff' now appear in both mailboxes, including `location-f`/`sync-target-f`'s mailbox.

```console
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:my-org:wmw-s
Current workspace is "root:my-org:wmw-s".
ubuntu@ip-172-31-31-125:~/edge-mc$ ky sps edge-placement-s      
apiVersion: edge.kcp.io/v1alpha1
destinations:
- cluster: 1gcc1zbbgm94lod9
  locationName: location-g
  syncTargetName: sync-target-g
  syncTargetUID: ca3eb1cd-a8ac-4b4a-85b7-0be352500631
kind: SinglePlacementSlice
metadata:
  annotations:
    kcp.io/cluster: 23qa6hllb50lvxjd
  creationTimestamp: "2023-04-20T03:26:08Z"
  generation: 1
  name: edge-placement-s
  ownerReferences:
  - apiVersion: edge.kcp.io/v1alpha1
    kind: EdgePlacement
    name: edge-placement-s
    uid: 4d015af9-61af-4a8f-b052-525e8ca777cd
  resourceVersion: "1255"
  uid: 781e5f45-f7da-4c0e-9da9-20bd81c93aa4
ubuntu@ip-172-31-31-125:~/edge-mc$ kc edit epl edge-placement-s
edgeplacement.edge.kcp.io/edge-placement-s edited
ubuntu@ip-172-31-31-125:~/edge-mc$ ky sps edge-placement-s      
apiVersion: edge.kcp.io/v1alpha1
destinations:
- cluster: 1gcc1zbbgm94lod9
  locationName: location-f
  syncTargetName: sync-target-f
  syncTargetUID: a113b61e-83ed-4eac-86cc-73f114d79e4c
- cluster: 1gcc1zbbgm94lod9
  locationName: location-g
  syncTargetName: sync-target-g
  syncTargetUID: ca3eb1cd-a8ac-4b4a-85b7-0be352500631
kind: SinglePlacementSlice
metadata:
  annotations:
    kcp.io/cluster: 23qa6hllb50lvxjd
  creationTimestamp: "2023-04-20T03:26:08Z"
  generation: 2
  name: edge-placement-s
  ownerReferences:
  - apiVersion: edge.kcp.io/v1alpha1
    kind: EdgePlacement
    name: edge-placement-s
    uid: 4d015af9-61af-4a8f-b052-525e8ca777cd
  resourceVersion: "1301"
  uid: 781e5f45-f7da-4c0e-9da9-20bd81c93aa4
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c
Current workspace is "root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c".
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME          STATUS   AGE
commonstuff   Active   7m32s
default       Active   13m
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get deployment -A
NAMESPACE     NAME      READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff   commond   0/0     0            0           7m39s
```

#### Delete an EdgePlacement

In these steps, I delete `edge-placement-c`.
It's expected that 'commonstuff' now disappear in both mailboxes.

```console
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:my-org:wmw-c
Current workspace is "root:my-org:wmw-c".
ubuntu@ip-172-31-31-125:~/edge-mc$ kc delete epl --all
edgeplacement.edge.kcp.io "edge-placement-c" deleted
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get sps
No resources found
ubuntu@ip-172-31-31-125:~/edge-mc$ kubectl ws root:espw:1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631
Current workspace is "root:espw:1gcc1zbbgm94lod9-mb-ca3eb1cd-a8ac-4b4a-85b7-0be352500631".
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME           STATUS   AGE
commonstuff    Active   8m59s
default        Active   14m
specialstuff   Active   8m59s
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get deployment -A
NAMESPACE      NAME       READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff    commond    0/0     0            0           9m4s
specialstuff   speciald   0/0     0            0           9m4s
ubuntu@ip-172-31-31-125:~/edge-mc$ date
Thu Apr 20 03:36:50 UTC 2023
ubuntu@ip-172-31-31-125:~/edge-mc$ date
Thu Apr 20 03:38:41 UTC 2023
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME           STATUS   AGE
commonstuff    Active   11m
default        Active   17m
specialstuff   Active   11m
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get deployment -A
NAMESPACE      NAME       READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff    commond    0/0     0            0           11m
specialstuff   speciald   0/0     0            0           11m
ubuntu@ip-172-31-31-125:~/edge-mc$ tail placement-translator.log 
W0420 03:27:21.644035  691389 shared_informer.go:200] The sharedIndexInformer has started, run more than once is not allowed
I0420 03:27:21.644800  691389 map-types.go:342] "Put" map="what" key="23qa6hllb50lvxjd:edge-placement-s" val={Downsync:map[{APIGroup: Resource:namespaces Name:specialstuff}:{APIVersion:v1 IncludeNamespaceObject:false} {APIGroup:apis.kcp.io Resource:apibindings Name:bind-kube}:{APIVersion:v1alpha1 IncludeNamespaceObject:false}] Upsync:[{APIGroup:group1.test Resources:[sprockets flanges] Namespaces:[orbital] Names:[george cosmo]} {APIGroup:group3.test Resources:[widgets] Namespaces:[] Names:[*]}]}
E0420 03:27:21.694506  691389 workload-projector.go:604] "Failed to create SyncerConfig" err="syncerconfigs.edge.kcp.io \"the-one\" already exists" worker=0 syncerConfig={Cluster:pgy6ew6j1mz900sj Name:the-one} destination={Cluster:1gcc1zbbgm94lod9 LocationName:location-g SyncTargetName:sync-target-g SyncTargetUID:ca3eb1cd-a8ac-4b4a-85b7-0be352500631}
I0420 03:27:21.744504  691389 shared_informer.go:289] Caches are synced for workload-projector
I0420 03:27:21.744504  691389 shared_informer.go:289] Caches are synced for workload-projector
I0420 03:27:22.721282  691389 request.go:601] Waited for 1.049886149s due to client-side throttling, not priority and fairness, request: GET:https://172.31.31.125:6443/clusters/18hvxcwlfhako08h/apis/networking.k8s.io/v1/ingresses?limit=500&resourceVersion=0
E0420 03:27:22.728784  691389 workload-projector.go:803] "Failed to create object in mailbox workspace" err="apibindings.apis.kcp.io \"bind-kube\" already exists" worker=3 objectRef={cluster:23qa6hllb50lvxjd groupResource:{Group:apis.kcp.io Resource:apibindings} namespace:no NS name:bind-kube} destination={Cluster:1gcc1zbbgm94lod9 LocationName:location-g SyncTargetName:sync-target-g SyncTargetUID:ca3eb1cd-a8ac-4b4a-85b7-0be352500631}
I0420 03:34:11.457189  691389 map-types.go:342] "Put" map="where" key="23qa6hllb50lvxjd:edge-placement-s" val="[&{{ } {edge-placement-s    781e5f45-f7da-4c0e-9da9-20bd81c93aa4 1301 2 2023-04-20 03:26:08 +0000 UTC <nil> <nil> map[] map[kcp.io/cluster:23qa6hllb50lvxjd] [{edge.kcp.io/v1alpha1 EdgePlacement edge-placement-s 4d015af9-61af-4a8f-b052-525e8ca777cd <nil> <nil>}] []  [{scheduler Update edge.kcp.io/v1alpha1 2023-04-20 03:26:08 +0000 UTC FieldsV1 {\"f:destinations\":{},\"f:metadata\":{\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"4d015af9-61af-4a8f-b052-525e8ca777cd\\\"}\":{}}}} }]} [{1gcc1zbbgm94lod9 location-f sync-target-f a113b61e-83ed-4eac-86cc-73f114d79e4c} {1gcc1zbbgm94lod9 location-g sync-target-g ca3eb1cd-a8ac-4b4a-85b7-0be352500631}]}]"
I0420 03:35:49.011590  691389 map-types.go:342] "Put" map="what" key="18hvxcwlfhako08h:edge-placement-c" val={Downsync:map[] Upsync:[]}
I0420 03:35:49.024857  691389 map-types.go:346] "Delete" map="where" key="18hvxcwlfhako08h:edge-placement-c"
ubuntu@ip-172-31-31-125:~/edge-mc$ kc ws root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c
Current workspace is "root:espw:1gcc1zbbgm94lod9-mb-a113b61e-83ed-4eac-86cc-73f114d79e4c".
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns
NAME          STATUS   AGE
commonstuff   Active   39m
default       Active   45m
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get deployment -A
NAMESPACE     NAME      READY   UP-TO-DATE   AVAILABLE   AGE
commonstuff   commond   0/0     0            0           39m
```

### Steps To Reproduce

See the description section.

### Expected Behaviour

See the description section.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
## Summary
This PR changes the placement translator to remedy the oversights noted in issue #320.

It was not reacting to workload objects in the mailbox workspace regardless of whether they are desired, so not all undesired ones were not being deleted.

It was not reconsidering all the source objects when the projection criteria change, so consequent deltas were not happening.

## Related issue(s)

Fixes #320 

/cc @waltforme 
/cc @yana1205 
/cc @francostellari

### CODE REVIEW

**Comment by waltforme** [comment]
I tested this PR using [the manifests of the turbo demo](https://github.com/edge-experiments/gitops-source/tree/main/edge-mc). It almost works with a tiny thing.

I changed the `locationSelectors` of my edgeplacement, trying to move workload from cluster1 to cluster2. 

What works:
- The syncer config in cluster1's mailbox got correctly emptied;
- The Deployment object in cluster1's mailbox got correctly deleted.

The tiny thing:
- The namespace which contained the deleted Deployment remained in cluster1's mailbox. (and vice versa: I changed back to cluster1, the namespace in cluster2's mailbox was still there)

I checked the namespace's metadata and saw the newly introduced label `edge.kcp.io/projected: "yes"`. So I guess the namespace is also part of the managed workload, and it is expected that this namespace also got deleted, right?

```
ubuntu@ip-172-31-31-125:~/edge-mc$ kc get ns optimized -oyaml | yq .metadata
annotations:
  kcp.io/cluster: 3597eq3gqxel82mx
creationTimestamp: "2023-04-21T15:34:18Z"
labels:
  edge.kcp.io/projected: "yes"
  kubernetes.io/metadata.name: optimized
name: optimized
resourceVersion: "3047"
uid: 20264998-b42e-43fb-92e3-a93307af5fee
```

I think the namespace thing doesn't affect the turbo demo. So
/lgtm

**Comment by MikeSpreitzer** [comment]
The namespace thing is a difficulty that is noted in the code; you should see a comment at the place where a deletion might happen but doesn't because the code can not be sure there isn't another reason to keep the namespace.  I meant to mention it in the doc too but forgot.  Will add that in a follow-on.

/approve

File: pkg/placement/workload-projector.go [hunk 1]
```diff
@@ -364,41 +368,99 @@ func (wp *workloadProjector) newPerDestinationLocked(destination SinglePlacement
...
+	} else {
+		sgvr := MetaGroupResourceToSchema(gr).WithVersion(apiVersion)
+		wpd.logger.V(4).Info("Creating informer at destination", "groupResource", gr, "apiVersion", apiVersion, "namespaced", namespaced)
+		duo = dynamicDuo{
+			apiVersion: apiVersion,
+			namespaced: namespaced,
+			client:     wpd.dynamicClient.Resource(sgvr)}
+		if mgrIsNamespace(gr) {
+			// No way to know if a namespace is needed for other reasons,
+			// so no point in reacting to them.
```
Comment (MikeSpreitzer) [inline pkg/placement/workload-projector.go:444]: @waltforme : this is the comment I mentioned.

Comment (waltforme) [reply pkg/placement/workload-projector.go:444]: Oh I see. I was testing the PR by running it, but not have enough bandwidth to review the code in full. Thanks for the info!

**Pull request merged**